File: 2023/d6/d6.py
import numpy as np

# reading the input
with open('./d6_input.txt') as f:  
    lines = [lines.rstrip() for lines in f]

#### Part 1

time = np.array(lines[0].split(':')[1].split(), dtype=np.int64)
dist = np.array(lines[1].split(':')[1].split(), dtype=np.int64)

ways_to_win = []

for time, dist in zip(time, dist):
    time_arrays = np.arange(0, time+1)
    results = [x for x in time_arrays if (time - x) * x > dist]
    ways_to_win.append(len(results))

print(np.prod(ways_to_win))


#### Part 2

time = np.array(int(lines[0].split(':')[1].replace(" ","")), dtype=np.int64)
dist = np.array(int(lines[1].split(':')[1].replace(" ","")), dtype=np.int64)

time_arrays = np.arange(0, time+1)
results = [x for x in time_arrays if (time - x) * x > dist]

print(len(results))

File: 2023/d06/d6.jl
# reading the input
lines = readlines("./d6_input.txt")
lines = rstrip.(lines)  # Julia uses broadcasting ('.') for element-wise operations

#### Part 1

time = parse.(Int64, split(split(lines[1], ':')[2]))
dist = parse.(Int64, split(split(lines[2], ':')[2]))

ways_to_win = Int[]

for (t, d) in zip(time, dist)
    time_arrays = 0:t
    results = [x for x in time_arrays if (t - x) * x > d]
    push!(ways_to_win, length(results))
end

println(prod(ways_to_win))

#### Part 2

time_new = parse(Int64, replace(split(lines[1], ':')[2], " " => ""))
dist_new = parse(Int64, replace(split(lines[2], ':')[2], " " => ""))

time_arrays = 0:time_new
results = [x for x in time_arrays if (time_new - x) * x > dist_new]

println(length(results))
